tapchannel/aux_funding.go

```diff
@@ -0,0 +1,933 @@
+	}
+
+	fundingTxid := finalFundingTx.FinalTx.TxHash()
+
+	return &fundingTxid, nil
+}
+
+// chanFunder is the main event loop that controls the asset specific portions
+// of the funding request.
+func (f *FundingController) chanFunder() {
+	defer f.wg.Done()
+
+	fundingFlows := make(fundingFlowIndex)
+
+	for {
+		select {
+
+		// A new funding request has arrived. We'll set up the funding
+		// state, send our input proofs, then kick off the channel
+		// funding asynchronously.
+		case fundReq := <-f.newFundingReqs:
+			// To start, we'll make a new pending asset funding
+			// desc. This'll be our scratch pad during the asset
+			// funding process.
+			tempPID, err := newPendingChanID()
+			if err != nil {
+				fmt.Printf("unable to create new pending "+
+					"chan ID: %v", err)
+
+				fundReq.errChan <- err
+				continue
+			}
+			fundingState := &pendingAssetFunding{
+				peerPub:   fundReq.PeerPub,
+				pid:       tempPID,
+				initiator: true,
+				assetID:   fundReq.AssetID,
+				amt:       fundReq.Amt,
+				feeRate:   fundReq.FeeRate,
+			}
+
+			fundingFlows[tempPID] = fundingState
+
+			// With our initial state created, we'll now attempt to
+			// fund the channel on the TAP level with a vPacket.
+			fundingVpkt, err := f.fundVpkt(
+				fundReq.ctx, fundReq.AssetID, fundReq.Amt,
+			)
+			if err != nil {
+				fErr := fmt.Errorf("unable to fund "+
+					"vPacket: %v", err)
+				fundReq.errChan <- fErr
+				continue
+			}
+
+			// Before we can send our OpenChannel message, we'll
+			// need to derive then send a series of ownership
+			// proofs to the remote party.
+			err = f.sendInputOwnershipProofs(
+				fundReq.PeerPub, fundingVpkt.VPacket,
+				fundingState,
+			)
+			if err != nil {
+				fErr := fmt.Errorf("unable to send input "+
+					"ownership proofs: %v", err)
+				fundReq.errChan <- fErr
+				continue
+			}
+
+			// With the ownership proof sent, we'll now spawn a
+			// goroutine to take care of the final funding steps.
+			f.wg.Add(1)
+			go func() {
+				defer f.wg.Done()
+				fundingTxid, err := f.completeChannelFunding(
+					fundReq.ctx, fundingState, fundingVpkt,
+				)
+				if err != nil {
+					fundReq.errChan <- err
+					return
+				}
+
+				fundReq.respChan <- fundingTxid
+			}()
+
+		// The remote party has sent us some upfront proof for channel
+		// asset inputs. We'll log this pending chan ID, then validate
+		// the proofs included.
+		case msg := <-f.msgs:
+			// A new proof message has just come in, so we'll
+			// extract the real proof wire message from the opaque
+			// message.
+			assetProof, assetFunding := fundingFlows.fromMsg(msg)
+
+			tempPID := assetProof.TempChanID.Val
+
+			// TODO(rosabeef): verify that has challenge witness
+			// before?
+
+			// Next, we'll validate this proof to make sure that
+			// the initiator is actually able to spend these
+			// outputs in the funding transaction.
+			ctxb := context.Background()
+			_, err := assetProof.Proof.Val.Verify(
```

Comment:
Oh, I just noticed something: In `Verify` we only check _either_ the challenge witness _or_ the state transition witness, even if both is present. I guess we should change that, otherwise I could smuggle in an asset that I can sign for but that was never authorized to belong to me.

Reply:
So here I always pass in `nil` for the prev asset, which _should_ force just the challenge witness validation. Can you give a more complete example of the smuggle edge case?

Reply:
Discussed offline, we'll check the full asset provenance at another point, here we just need to make sure the remote party can sign for the assets.